pkg/sql/logictest/testdata/logic_test/composite_types

```diff
@@ -0,0 +1,213 @@
+#
+#statement ok
+#INSERT INTO tab2 VALUES(((1, 2), (3, 4)))
+#
+#query TTII
+#SELECT a, (a).t1, ((a).t1).a, ((a).t1).b FROM tab2
+#----
+#("(1,2)","(3,4)")  (1,2)  1  2
+#
+## Can't drop type t because tab, tab2, and t2 depend on it
+#statement error cannot drop type \"t\" because other objects .* still depend on it
+#DROP TYPE t
+#
+## Can't drop type t2 because tab2 depends on it
+#statement error cannot drop type \"t2\" because other objects .* still depend on it
+#DROP TYPE t2
+#
+#statement ok
+#DROP TABLE tab2
+
+statement ok
+DROP TABLE tab
+
+query TTTT
+SELECT database_name, schema_name, descriptor_name, create_statement FROM crdb_internal.create_type_statements
+----
+test  public  t   CREATE TYPE public.t AS (a INT8, b INT8)
+
+## Can't drop type t because t2 depends on it
+#statement error cannot drop type \"t\" because other objects .* still depend on it
+#DROP TYPE t
+#
+#statement ok
+#DROP TYPE t2
+
+statement ok
+DROP TYPE t
+
+# An empty type is valid.
+statement ok
+CREATE TYPE t AS ()
+
+statement ok
+DROP TYPE t
+
+# Composite types which reference other types of UDTs are not yet supported.
+statement ok
+CREATE TYPE e AS ENUM ('a', 'b', 'c')
+
+# We'll use tab to check the implicit table alias type.
+statement ok
+CREATE TABLE tab (a INT, b INT)
+
+statement error composite types that reference user-defined types not yet supported
+CREATE TYPE t AS (e e)
+
+# This should fail - we shouldn't persist implicit table types.
+statement error composite types that reference user-defined types not yet supported
+CREATE TYPE t AS (a tab)
+
+statement error composite types that reference user-defined types not yet supported
+CREATE TYPE t AS (a pg_catalog.pg_class)
+
+# Test that if an composite type value is being used by a default expression or
+# computed column, we disallow dropping it.
+subtest drop_used_composite_type_values
+
+statement ok
+CREATE TYPE t AS (a INT, b TEXT)
+
+statement ok
+CREATE TABLE a (a INT DEFAULT (((1, 'hi')::t).a))
+
+statement error cannot drop type \"t\" because other objects .* still depend on it
+DROP TYPE t
+
+# Re-enable when #91972 is resolved.
+#statement ok
+#ALTER TABLE a ALTER COLUMN a SET DEFAULT 3
+#
+#statement ok
+#DROP TYPE t;
+#CREATE TYPE t AS (a INT, b TEXT);
+
+statement ok
+DROP TABLE a;
+CREATE TABLE a (a INT ON UPDATE (((1, 'hi')::t).a))
+
+statement error cannot drop type \"t\" because other objects .* still depend on it
+DROP TYPE t
+
+# Re-enable when #91972 is resolved.
+#statement ok
+#ALTER TABLE a ALTER COLUMN a SET ON UPDATE 3
+#
+#statement ok
+#DROP TYPE t;
+#CREATE TYPE t AS (a INT, b TEXT);
+
+statement ok
+DROP TABLE a;
+CREATE TABLE a (a INT AS (((1, 'hi')::t).a) STORED)
+
+statement error cannot drop type \"t\" because other objects .* still depend on it
+DROP TYPE t
+
+# Re-enable when #91972 is resolved.
+#statement ok
+#ALTER TABLE a ALTER COLUMN a DROP STORED
+#
+#statement ok
+#DROP TYPE t;
+#CREATE TYPE t AS (a INT, b TEXT);
```

Comment:
Let's make a note in #91972 to remember to re-enable this.

Reply:
Done.